### ISSUE TITLE
ansible-container cannot cope with requirements upgrade

### ISSUE THREAD

**Issue by cans** [ansible/ansible-container#890]
##### ISSUE TYPE

 - Bug Report

##### container.yml 

```yaml
version: "2"
settings:
  conductor:
    base: debian:9
  project_name: ansible-testing
  deployment_output_path: ./ansible-deployment
services:
  base:
    from: "python:3.6-slim-stretch"
    roles:
      - role: "cans.package-install"
        pkginstall_packages:
          # - python3.6
          - bash
          - dumb-init
          - sudo
      - role: "cans.user-make"
        usermake_upload_ssh_key_to_target: false
        usermake_users:
          - name: "admin"
            sudoer: true
            passwordless_sudo: true
          - name: "regular"
            sudoer: false
    entrypoint: /usr/bin/dumb-init
registries: {}
```

##### OS / ENVIRONMENT

```
Ansible Container, version 0.9.2
Linux, cornil-weus, 4.14.0-3-amd64, #1 SMP Debian 4.14.13-1 (2018-01-14), x86_64
3.6.4 (default, Jan  5 2018, 02:13:53) 
[GCC 7.2.0] /home/ncaniart/.virtualenvs/cans-containers/bin/python3.6
{
  "ID": "NT5T:LRIH:V2OX:T4DS:ROOU:2NJJ:TDFG:NRLQ:EKFO:CH2N:SCEM:YCR4",
  "Containers": 1,
  "ContainersRunning": 0,
  "ContainersPaused": 0,
  "ContainersStopped": 1,
  "Images": 13,
  "Driver": "overlay2",
  "DriverStatus": [
    [
      "Backing Filesystem",
      "extfs"
    ],
    [
      "Supports d_type",
      "true"
    ],
    [
      "Native Overlay Diff",
      "true"
    ]
  ],
  "SystemStatus": null,
  "Plugins": {
    "Volume": [
      "local"
    ],
    "Network": [
      "bridge",
      "host",
      "macvlan",
      "null",
      "overlay"
    ],
    "Authorization": null,
    "Log": [
      "awslogs",
      "fluentd",
      "gcplogs",
      "gelf",
      "journald",
      "json-file",
      "logentries",
      "splunk",
      "syslog"
    ]
  },
  "MemoryLimit": true,
  "SwapLimit": true,
  "KernelMemory": true,
  "CpuCfsPeriod": true,
  "CpuCfsQuota": true,
  "CPUShares": true,
  "CPUSet": true,
  "IPv4Forwarding": true,
  "BridgeNfIptables": true,
  "BridgeNfIp6tables": true,
  "Debug": false,
  "NFd": 20,
  "OomKillDisable": true,
  "NGoroutines": 34,
  "SystemTime": "2018-02-18T20:11:19.669468492+01:00",
  "LoggingDriver": "json-file",
  "CgroupDriver": "cgroupfs",
  "NEventsListener": 0,
  "KernelVersion": "4.14.0-3-amd64",
  "OperatingSystem": "Debian GNU/Linux buster/sid",
  "OSType": "linux",
  "Architecture": "x86_64",
  "IndexServerAddress": "https://index.docker.io/v1/",
  "RegistryConfig": {
    "AllowNondistributableArtifactsCIDRs": [],
    "AllowNondistributableArtifactsHostnames": [],
    "InsecureRegistryCIDRs": [
      "127.0.0.0/8"
    ],
    "IndexConfigs": {
      "docker.io": {
        "Name": "docker.io",
        "Mirrors": [],
        "Secure": true,
        "Official": true
      }
    },
    "Mirrors": []
  },
  "NCPU": 4,
  "MemTotal": 8236167168,
  "GenericResources": null,
  "DockerRootDir": "/var/lib/docker",
  "HttpProxy": "",
  "HttpsProxy": "",
  "NoProxy": "",
  "Name": "cornil-weus",
  "Labels": [],
  "ExperimentalBuild": false,
  "ServerVersion": "17.12.0-ce",
  "ClusterStore": "",
  "ClusterAdvertise": "",
  "Runtimes": {
    "runc": {
      "path": "docker-runc"
    }
  },
  "DefaultRuntime": "runc",
  "Swarm": {
    "NodeID": "",
    "NodeAddr": "",
    "LocalNodeState": "inactive",
    "ControlAvailable": false,
    "Error": "",
    "RemoteManagers": null
  },
  "LiveRestoreEnabled": false,
  "Isolation": "",
  "InitBinary": "docker-init",
  "ContainerdCommit": {
    "ID": "89623f28b87a6004d4b785663257362d1658a729",
    "Expected": "89623f28b87a6004d4b785663257362d1658a729"
  },
  "RuncCommit": {
    "ID": "b2567b37d7b75eb4cf325b77297b140ea686ce8f",
    "Expected": "b2567b37d7b75eb4cf325b77297b140ea686ce8f"
  },
  "InitCommit": {
    "ID": "949e6fa",
    "Expected": "949e6fa"
  },
  "SecurityOptions": [
    "name=apparmor",
    "name=seccomp,profile=default"
  ]
}
{
  "Platform": {
    "Name": ""
  },
  "Components": [
    {
      "Name": "Engine",
      "Version": "17.12.0-ce",
      "Details": {
        "ApiVersion": "1.35",
        "Arch": "amd64",
        "BuildTime": "2017-12-27T20:09:29.000000000+00:00",
        "Experimental": "false",
        "GitCommit": "c97c6d6",
        "GoVersion": "go1.9.2",
        "KernelVersion": "4.14.0-3-amd64",
        "MinAPIVersion": "1.12",
        "Os": "linux"
      }
    }
  ],
  "Version": "17.12.0-ce",
  "ApiVersion": "1.35",
  "MinAPIVersion": "1.12",
  "GitCommit": "c97c6d6",
  "GoVersion": "go1.9.2",
  "Os": "linux",
  "Arch": "amd64",
  "KernelVersion": "4.14.0-3-amd64",
  "BuildTime": "2017-12-27T20:09:29.000000000+00:00"
}
```

##### SUMMARY

When one updates its requirements (newer version of a role), ansible-container cannot seem to cope with upgrading those requirements. When using plain ansible one has the option to force reinstalling roles with the `--force` command line option (everything not just what's no longer up-to-date). In this particular case, I couldn't find a way to ask the conductor to do the same, I found no such option for the build command. Having the ability to forward the `--force` option somehow, seems reasonable.

Role upgrade is a bit of a pain whether you use plain ansible or ansible-container, and I would say that part of this issue weighs on ansible-galaxy's shoulders.

##### STEPS TO REPRODUCE
Use the above `container.yml` with the following requirements.yml.
```yaml
- src: cans.user-make
  version: 'v1.1.0'

- src: cans.package-install
  version: 'v2.0.0'
```
Run: `ansible-container build`. The build will fail because of a bug in the `cans.user-make` role (that's my fault, not yours). Change the requirements for:

```yaml
- src: cans.user-make
  version: 'v1.1.1'

- src: cans.package-install
  version: 'v2.0.0'
```
Run `ansible-container build` again.

##### EXPECTED RESULTS

I would expect the requirements to be properly upgraded and the build to complete successfully.

##### ACTUAL RESULTS

The build fails with the error:
```
Traceback (most recent call last):
  File "/usr/local/bin/conductor", line 11, in <module>
    load_entry_point('ansible-container', 'console_scripts', 'conductor')()
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/cli.py", line 389, in conductor_commandline
    conductor_config = AnsibleContainerConductorConfig(list_to_ordereddict(containers_config))
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/config.py", line 297, in __init__
    self._process_services()
  File "/_ansible/container/config.py", line 357, in _process_services
    role_metadata = get_metadata_from_role(role_name)
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/utils/__init__.py", line 275, in get_metadata_from_role
    return get_content_from_role(role_name, os.path.join('meta', 'container.yml'))
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/utils/__init__.py", line 264, in get_content_from_role
    role_path = resolve_role_to_path(role_name)
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/utils/__init__.py", line 210, in resolve_role_to_path
    loader=loader)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/role/include.py", line 59, in load
    return ri.load_data(data, variable_manager=variable_manager, loader=loader)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/base.py", line 244, in load_data
    ds = self.preprocess_data(ds)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/role/definition.py", line 94, in preprocess_data
    (role_name, role_path) = self._load_role_path(role_name)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/role/definition.py", line 187, in _load_role_path
    raise AnsibleError("the role '%s' was not found in %s" % (role_name, ":".join(role_search_paths)), obj=self._ds)
ansible.errors.AnsibleError: the role 'cans.user-make' was not found in ./roles:/src/roles:/etc/ansible/roles:.
```
[The complete debug output](https://gist.github.com/cans/8adaf0771cf148c2786eaaf5f039c9c6) I tried to get rid of the current images with `ansible-container destroy` but it failed in the exact same way: 

```
ansible-container destroy
Parsing conductor CLI args.
Traceback (most recent call last):
  File "/usr/local/bin/conductor", line 11, in <module>
    load_entry_point('ansible-container', 'console_scripts', 'conductor')()
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/cli.py", line 389, in conductor_commandline
    conductor_config = AnsibleContainerConductorConfig(list_to_ordereddict(containers_config))
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/config.py", line 297, in __init__
    self._process_services()
  File "/_ansible/container/config.py", line 357, in _process_services
    role_metadata = get_metadata_from_role(role_name)
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/utils/__init__.py", line 275, in get_metadata_from_role
    return get_content_from_role(role_name, os.path.join('meta', 'container.yml'))
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/utils/__init__.py", line 264, in get_content_from_role
    role_path = resolve_role_to_path(role_name)
  File "/_ansible/container/__init__.py", line 19, in __wrapped__
    return fn(*args, **kwargs)
  File "/_ansible/container/utils/__init__.py", line 210, in resolve_role_to_path
    loader=loader)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/role/include.py", line 59, in load
    return ri.load_data(data, variable_manager=variable_manager, loader=loader)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/base.py", line 244, in load_data
    ds = self.preprocess_data(ds)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/role/definition.py", line 94, in preprocess_data
    (role_name, role_path) = self._load_role_path(role_name)
  File "/usr/local/lib/python2.7/dist-packages/ansible/playbook/role/definition.py", line 187, in _load_role_path
    raise AnsibleError("the role '%s' was not found in %s" % (role_name, ":".join(role_search_paths)), obj=self._ds)
ansible.errors.AnsibleError: the role 'cans.user-make' was not found in ./roles:/src/roles:/etc/ansible/roles:.
Conductor terminated. Cleaning up.	command_rc=1 conductor_id=e1ea4d8679d24823ca248ffe567f6ad877e3d4a8a5733d1dca30c74f80d9c7e4 save_container=False
ERROR	Conductor exited with status 1	
```

The only solution I found to get out of this, was to delete everything using `docker rm` and `docker rmi` commands.

**Comment by j00bar** [comment]
Thank you for your report! In line 154 of your gist, Docker reports using cache for the Conductor, which if your `requirements.yml` had changed, it wouldn't. Any thoughts?

**Comment by cans** [comment]
Hi @j00bar !

Sorry for the lag. So, I tried to reproduce the issue again. Took me a while, but I actually managed to reproduce it once... when I sent vim in the background before actually saving the modified `requirements.yml`, then ran ansible-container build... :confused: :

Apologies for your wasted time. Hopefully having that written down here will prevent further waste in the future. And I'll just close this.
